src/action.py

```diff
@@ -1,30 +1,67 @@
 
 import glob
 import click
-import os
+import re
+import config
+
+from sys import exit
+from os import listdir
+from os.path import join, dirname, relpath, isfile, isdir
 from models.config import Config
 from collections import defaultdict
 from itertools import chain
 
 class ClickGlob:
-    def __glob_configs(namespace):
-        parts = ['/var/lib/adminware/tools',
-                 namespace, '*/config.yaml']
-        paths = glob.glob(os.path.join(*parts))
-        return list(map(lambda x: Config(x), paths))
+    def __glob_dirs(namespace):
+        new_dir = config.join_with_tool_location(namespace)
+        if isdir(new_dir):
+            parts = listdir(new_dir)
+            paths = list(map(lambda x: join(new_dir, x), parts))
+            return paths
+        else:
+            return []
+
+    # repeats __glob_dirs to get all valid paths (ending in config.yaml) at once
+    def __glob_all(namespace):
+        def __glob_recur(namespace, all_paths=[]):
+            paths = ClickGlob.__glob_dirs(namespace)
+            for path in paths:
+                if isfile(ClickGlob.join_config(path)):
+                    all_paths += [ClickGlob.join_config(path)]
+                else:
+                    namespace = path[len(config.LEADER + config.TOOL_LOCATION):]
+                    __glob_recur(namespace, all_paths)
+            return all_paths
+
+        return __glob_recur(namespace)
+
 
     def command(click_group, namespace):
         # command_func is either run_open or run_batch, what this is decorating
         def __command(command_func):
-            actions = list(map(lambda x: Action(x), ClickGlob.__glob_configs(namespace)))
-            for action in actions:
-                action.create(click_group, command_func)
+            __command_helper(click_group, namespace, command_func)
+
+        def __command_helper(cur_group, cur_namespace, command_func):
+            paths = ClickGlob.__glob_dirs(cur_namespace) # will generate a list of paths at level 'namespace'
```

Comment:
Is it possible to use `ClickGlob.__glob_all` here? Currently this class re-implements the loop that is used within `glob_all`, however it also adds an additional condition check.

This means any method that relies on `glob_all` will need to preform this same condition check or risk being inconsistent.

Also functions that are prefixed with `__` are private and should not be called from outside of the class.

Reply:
Yep I did notice the shared recursion expanding the namespace with each iteration, however I quietly decided against any attempt to combine the two due to:
- the difference in their parameters (one needs to take the `command_func` itself and the current `group` while the other does not)
- the difference in their 'if' blocks (one creates an `Action` object & a `click.command` while the other just forms a list) 
- the difference in their else blocks (one creates a new group while the other only calls recursively. 

In short the only this they share is method of recursion 

To me these differences were too irreconcilable to attempt DRYing up while keeping any semblance of comprehension, what's your opinion?

Reply:
Ideally they would share a common recursion method as it is technically possible for the `batch family` commands to pick up more configs than `batch`. However DRYing it up is going to be a bit of a pain.

Can you create an issue for this and reference this PR?

Reply:
Yeah, absolutely